ddl/partition.go

```diff
@@ -1206,3 +1208,31 @@ func truncateTableByReassignPartitionIDs(t *meta.Meta, tblInfo *model.TableInfo)
 	tblInfo.Partition.Definitions = newDefs
 	return nil
 }
+
+func onAlterTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
+	var rules []*placement.Rule
+	err := job.DecodeArgs(&rules)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return 0, errors.Trace(err)
+	}
+
+	ver, err := t.GetSchemaVersion()
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	ok, err := infosync.UpdatePlacementRules(nil, rules)
```

Comment:
Also check partition here. `AlterTablePartition` just checks legality before executing and put it into the job queue. When it is executed, the partition may be dropped or truncated.

Reply:
https://github.com/pingcap/tidb/pull/18611/commits/249cf8e4c463e29a6b5f71fdd9061cb8e1b4883f